DigitalLearningSolutions.Data/Services/DelegateApprovalsService.cs

```diff
@@ -1,34 +1,135 @@
+            ILogger<DelegateApprovalsService> logger
+        )
         {
             this.userDataService = userDataService;
             this.customPromptsService = customPromptsService;
+            this.emailService = emailService;
+            this.logger = logger;
         }
 
-        public List<(DelegateUser delegateUser, List<CustomPromptWithAnswer> prompts)> GetUnapprovedDelegatesWithCustomPromptAnswersForCentre(int centreId)
+        public List<(DelegateUser delegateUser, List<CustomPromptWithAnswer> prompts)>
+            GetUnapprovedDelegatesWithCustomPromptAnswersForCentre(int centreId)
         {
             var users = userDataService.GetUnapprovedDelegateUsersByCentreId(centreId);
             var usersWithPrompts =
                 customPromptsService.GetCentreCustomPromptsWithAnswersByCentreIdForDelegateUsers(centreId, users);
 
             return usersWithPrompts;
         }
+
+        public void ApproveDelegate(int delegateId, int centreId)
+        {
+            var delegateUser = userDataService.GetDelegateUserByIdFromCentre(delegateId, centreId);
+
+            if (delegateUser == null)
+            {
+                throw new UserAccountNotFoundException($"Delegate user id {delegateId} not found at centre id {centreId}.");
+            }
+
+            if (delegateUser.Approved)
+            {
+                logger.LogWarning($"Delegate user id {delegateId} already approved.");
+            }
+            else
+            {
+                userDataService.ApproveDelegateUsers(new[] { delegateId });
+
+                if (string.IsNullOrWhiteSpace(delegateUser.EmailAddress))
+                {
+                    LogNoEmailWarning(delegateUser.Id);
+                }
+                else
+                {
+                    var delegateApprovalEmail = GenerateDelegateApprovalEmail
+                        (delegateUser.CandidateNumber, delegateUser.EmailAddress);
+                    emailService.SendEmail(delegateApprovalEmail);
+                }
+            }
+        }
+
+        public void ApproveAllUnapprovedDelegatesForCentre(int centreId)
+        {
+            var delegateUsers = userDataService.GetUnapprovedDelegateUsersByCentreId(centreId);
+
+            userDataService.ApproveDelegateUsers(delegateUsers.Select(d => d.Id));
+
+            var approvalEmails = new List<Email>();
+            foreach (var delegateUser in delegateUsers)
+            {
+                if (string.IsNullOrWhiteSpace(delegateUser.EmailAddress))
+                {
+                    LogNoEmailWarning(delegateUser.Id);
+                }
+                else
+                {
+                    var delegateApprovalEmail = GenerateDelegateApprovalEmail
+                        (delegateUser.CandidateNumber, delegateUser.EmailAddress);
+                    approvalEmails.Add(delegateApprovalEmail);
+                }
+            }
+            emailService.SendEmails(approvalEmails);

```

Comment:
Might be worth pulling this into a private method "SendDelegateApprovalEmails" (+ in the single approval case). Then, the high level purpose of this method gets clearer to read :)